CreditManagement/models.py

```diff
@@ -1,124 +1,103 @@
-    source_association = models.ForeignKey(Association, related_name="transaction_source", on_delete=models.SET_NULL, null=True, blank=True, verbose_name="The association giving the money")
-    amount = models.DecimalField(verbose_name="Money transferred", decimal_places=2, max_digits=4, validators=[MinValueValidator(Decimal('0.01'))])
-    target_user = models.ForeignKey(User, related_name="transaction_target", on_delete=models.SET_NULL, null=True, blank=True, verbose_name="The user receiving the money")
-    target_association = models.ForeignKey(Association, related_name="transaction_target", on_delete=models.SET_NULL, null=True, blank=True, verbose_name="The association recieving the money")
-    description = models.CharField(default="", blank=True, max_length=50)
+    """
+    Todo: the following database constraints should be in place:
+
+    CHECK(amount > 0),
+    CHECK(source_user IS NULL OR source_association IS NULL), -- there must be at most one source
+    CHECK(target_user IS NULL OR target_association IS NULL), -- there must be at most one target
+    -- there must be at least a source or a target
+    CHECK(NOT(source_user IS NULL AND source_association IS NULL AND target_user IS NULL AND target_association IS NULL)),
+
+    These probably need to be inserted using custom migration files, however these are not yet in git.
+    """
+    moment = models.DateTimeField(auto_now_add=True)
+    source_user = models.ForeignKey(settings.AUTH_USER_MODEL, related_name="transaction_source",
+                                    on_delete=models.PROTECT, null=True, blank=True)
+    source_association = models.ForeignKey(Association, related_name="transaction_source", on_delete=models.PROTECT,
+                                           null=True, blank=True)
+    target_user = models.ForeignKey(settings.AUTH_USER_MODEL, related_name="transaction_target",
+                                    on_delete=models.PROTECT, null=True, blank=True)
+    target_association = models.ForeignKey(Association, related_name="transaction_target", on_delete=models.PROTECT,
+                                           null=True, blank=True)
+    amount = models.DecimalField(decimal_places=2, max_digits=16, validators=[MinValueValidator(Decimal('0.01'))])
+    notes = models.CharField(max_length=200, blank=True)
+
+    # Optional reference to the dining list that caused this transaction, for informational purposes.
+    dining_list = models.ForeignKey(DiningList, related_name='transactions', on_delete=models.PROTECT, null=True,
+                                    blank=True)
+
+    objects = TransactionManager()
 
     def source(self):
-        if self.source_association is None:
-            return self.source_user
-        else:
-            return self.source_association
+        """
+        Returns the transaction source which is a user or an association.
+        """
+        return self.source_user if self.source_user else self.source_association
```

Comment:
Source user and source association can not both be non-null. Can you recheck your comment?

Reply:
Why not? This way we can check who, on behalf of that association, made the wire order. If at some point an order is contested we can check who made the order and trace it to a person.

Reply:
Ah :ok_hand: you meant for auditing. I do think it can be confusing for maintainers to have both source association and source user set, when the actual transaction source is only the association and not the user. For this kind of auditing information about the transaction it is maybe better to use the notes field or add a field `created_by` or something(?).